jenkins_pipelines/environments/common/pipeline-build-validation-cleanup.groovy

```diff
@@ -0,0 +1,249 @@
+                if (params.clean_proxy) {
+                    tfResourcesToDelete.add('proxy')
+                }
+                if (params.clean_monitoring_server) {
+                    tfResourcesToDelete.add('monitoring-server')
+                }
+                if (params.clean_retail) {
+                    tfResourcesToDelete.add('retail')
+                }
+
+                // Join the resources into a comma-separated string if there are any to delete
+                String tfResourcesToDeleteArg = defaultResourcesToDelete.isEmpty() ? '' : "--tf-resources-to-delete ${defaultResourcesToDelete.join(' ')}"
+
+                // Execute Terracumber CLI to deploy the environment without clients
+                sh """
+                    ${environmentVars}
+                    set +x
+                    ./terracumber-cli ${commonParams} --logfile ${logFile} --init --sumaform-backend ${sumaform_backend} --use-tf-resource-cleaner --init --runstep provision ${tfResourcesToDeleteArg}
+                """
+            }
+
+            stage('Redeploy the environment with new client VMs') {
+
+                // Run Terracumber to deploy the environment
+                sh """
+                    ${environmentVars}
+                    set +x
+                    ./terracumber-cli ${commonParams} --logfile ${resultdirbuild}/sumaform.log --init --sumaform-backend ${sumaform_backend} --runstep provision
+                """
+            }
+
+            stage('Copy the new custom repository json file to controller') {
+                if (params.push_new_custom_repositories) {
+                    // Generate custom_repositories.json file in the workspace from the value passed by parameter
+                    if (params.custom_repositories?.trim()) {
+                        writeFile file: 'custom_repositories.json', text: params.custom_repositories, encoding: "UTF-8"
+                    }
+
+                    // Generate custom_repositories.json file in the workspace using a Python script - MI Identifiers passed by parameter
+                    if (params.mi_ids?.trim()) {
+                        node('manager-jenkins-node') {
+                            checkout scm
+                            res_python_script_ = sh(script: "python3 jenkins_pipelines/scripts/json_generator/maintenance_json_generator.py --mi_ids ${params.mi_ids}", returnStatus: true)
+                            echo "Build Validation JSON script return code:\n ${json_content}"
+                            if (res_python_script != 0) {
+                                error("MI IDs (${params.mi_ids}) passed by parameter are wrong (or already released)")
+                            }
+                        }
+                    }
+                    sh(script: "${SUSEManagerCleanerProgram} --url ${controllerHostname} --product_version ${product_version} --mode update_custom_repositories")
+                }
+            }
```

Comment:
Until that point, all the stages where inside the concept of "Clean-up".
But... I'm not sure if we should have this stage here, at least with the current pipeline name.

Reply:
Indeed it could be a solution. By default, the BVs are not updating the custom repo file. We can definitely move this part to BV pipeline. I will propose to keep it as is for now and move this stage to BV in a near future. https://github.com/SUSE/spacewalk/issues/26013